library/src/armory/export/criteria.py

```diff
@@ -0,0 +1,512 @@
+        # batch_idx is 0, 1, 3, etc.)
+        exporter = Exporter(criterion=not_satisfied(every_n_batches(3)))
+
+    Args:
+        criterion: Nested criterion for which unsatisfied samples are to be
+            exported
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx, batch):
+        res = criterion(chain_name, batch_idx, batch)
+        if not res:
+            return True
+        if type(res) is bool:
+            return False
+        return set(range(len(batch))).difference(res)
+
+    return _criterion
+
+
+def every_n_batches(n: int) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches all samples from every nth batch.
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import every_n_batches
+
+        # Exports every sample in every 5th batch (when batch_idx is 4, 9, 14, etc.)
+        exporter = Exporter(criterion=every_n_batches(5))
+
+    Args:
+        n: Interval at which to export batches based on the index of the batch
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx: int, batch):
+        if n == 0:
+            return False
+        return (batch_idx + 1) % n == 0
+
+    return _criterion
+
+
+def first_n_batches(n: int) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches all samples from the first n batches.
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import first_n_batches
+
+        # Exports every sample in the first 3 batches (when batch_idx is 0, 1, and 2)
+        exporter = Exporter(criterion=every_n_batches(3))
+
+    Args:
+        n: Number of batches to be exported
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx: int, batch):
+        if n == 0:
+            return False
+        return batch_idx < n
+
+    return _criterion
+
+
+def every_n_samples_of_batch(n: int) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches every nth sample in every batch.
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import every_n_samples_of_batch
+
+        # Exports every other sample in every batch (when sample_idx is 1, 3, 5, etc.)
+        exporter = Exporter(criterion=every_n_samples_of_batch(5))
+
+    Args:
+        n: Interval at which to export samples based on the index of the sample
+            within each batch
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx: int, batch):
+        if n == 0:
+            return False
+        return [
+            sample_idx for sample_idx in range(len(batch)) if (sample_idx + 1) % n == 0
+        ]
+
+    return _criterion
+
+
+def every_n_samples(n: int) -> Exporter.Criterion:
```

Comment:
This criterion (and others below) relies on a fixed numbering of the dataset samples. Is is worth making this explicit in the `DataLoader` classes by setting `shufffle=False` and specifying this in the doc comments?

Reply:
I don't think that it should be a requirement that shuffle be disabled in the data loader in order to use these criteria. While you'll certainly get different samples exported when you start shuffling the dataset, I think those are still valid use-cases. But I think adding a comment to the docs along the lines of "different samples will be exported if the dataset has been shuffled" is worthwhile.

Reply:
I can see that case being potentially confusing. I'll have to think about how to detect and warn for this case. The data loader is the only entity that knows the dataset is being shuffled. The exporter/criteria should check for it & warn, but they don't have access to the dataloader or have any knowledge about whether the data is shuffled. The `Engine` isn't really the appropriate place to put this responsibility, but it's the only place with access to all the involved objects.

This probably warrants a separate discussion about evaluation compatibility validation in general. This isn't the only area in which it's possible to set up an evaluation with incompatible settings/entities.